### PR TITLE
Added overloads for simplified func-based activity and orchestrator registration

### PR DESCRIPTION
**Pull request by cgillum**
# Summary
## What changed?
Added new overloads for func-based activity and orchestrator registration.

For example, instead of doing this all the time:

```csharp
static string SendEmail(TaskActivityContext context, string message)
{
    // ...
}

registry.AddActivityFunc<string, string>(nameof(SendEmail), SendEmail);
```

I can now do this:

```csharp
[DurableTask(nameof(SendEmail))] // <-- optional
static string SendEmail(TaskActivityContext context, string message)
{
    // ...
}

registry.AddActivityFunc<string, string>(SendEmail);
```

## Why is this change needed?
The current methods are verbose and require writing repetitive code. I found it quite annoying when creating samples.

## Issues / work items
There are no work items associated with this PR
---

# Project checklist
- [ ] Release notes are not required for the next release
  - [ ] Otherwise: Notes added to `release_notes.md`
- [x] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [x] Breaking change? (no)
  - [ ] If yes:
    - Impact:
    - Migration guidance:
---

# AI-assisted code disclosure (required)
## Was an AI tool used? (select one)
- [ ] No
- [ ] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [x] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot /w Claude Opus 4.5
- AI-assisted areas/files: I don't know what this means?
- What you changed after AI output: Nothing that I can specifically remember

AI verification (required if AI was used):
- [x] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing
## Automated tests
- Result: Passed

## Manual validation (only if runtime/behavior changed)
N/A

---

# Notes for reviewers
- N/A

### CODE REVIEW

File: misc/misc.csproj [hunk 1]
```diff
@@ -18,6 +18,7 @@
     <None Include="$(RepoRoot).gitignore" />
     <None Include="$(RepoRoot)azure-pipelines-release.yml" />
     <None Include="$(RepoRoot)Directory.Build.*" />
+    <None Include="$(RepoRoot)Directory.Packages.*" />
```
[the remark this answers is not in the document]
Comment (cgillum) [reply misc/misc.csproj:21]: I'm going to keep this in here since it's not worth creating a separate PR for.

File: src/Abstractions/Entities/TaskEntity.cs [hunk 2]
```diff
@@ -148,6 +148,12 @@ public abstract class TaskEntity<TState> : ITaskEntity
     /// <remarks>The default implementation uses <see cref="Activator.CreateInstance()"/>.</remarks>
     protected virtual TState InitializeState(TaskEntityOperation entityOperation)
     {
+        // Throw if TState is a string, since strings are immutable and don't support dynamic creation of new instances.
+        if (typeof(TState) == typeof(string))
+        {
+            throw new InvalidOperationException("Entity state cannot be a string. Use a class or struct instead.");
+        }
+
```
[the remark this answers is not in the document]
Comment (cgillum) [reply src/Abstractions/Entities/TaskEntity.cs:156]: This is something I ran into when doing some unrelated prototyping and wanted to create a fix while it was on the top of my mind. I could make it into a separate PR, but not sure if it's really worth it.

File: src/Abstractions/DurableTaskRegistry.Activities.cs [hunk 3]
```diff
@@ -219,4 +230,167 @@ public DurableTaskRegistry AddActivityFunc(TaskName name, Action<TaskActivityCon
...
+    /// </exception>
+    static TaskName GetActivityNameFromDelegate(Delegate @delegate)
+    {
+        MethodInfo method = @delegate.Method;
+
+        // Check for DurableTaskAttribute on the method
+        DurableTaskAttribute? attribute = method.GetCustomAttribute<DurableTaskAttribute>();
+        if (attribute?.Name.Name is not null and not "")
+        {
+            return attribute.Name;
```
[the remark this answers is not in the document]
Comment (cgillum) [reply src/Abstractions/DurableTaskRegistry.Activities.cs]: I don't really see how it's possible for `attribute` to ever be null here, assuming it's not shared state that can be accessed (and mutated) by multiple threads concurrently (something we're not doing).

File: src/Abstractions/Entities/TaskEntity.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (halspang) [reply src/Abstractions/Entities/TaskEntity.cs:156]: @cgillum - Does this end up being a breaking change? Could users have a string based entity beforehand or would it fail elsewhere in the code and this is just making the error more descriptive?

File: src/Abstractions/DurableTaskRegistry.Activities.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (halspang) [reply src/Abstractions/DurableTaskRegistry.Activities.cs]: I agree it shouldn't be null here, but I would prefer if we kept with the `!string.IsNullOrEmpty(attribute?.Name.Name)` approach.  It's more consistent with other places in the code and, I think, easier to read.

File: src/Abstractions/DurableTaskRegistry.Orchestrators.cs [hunk 4]
```diff
@@ -220,4 +232,171 @@ public DurableTaskRegistry AddOrchestratorFunc(TaskName name, Action<TaskOrchest
...
+    /// <exception cref="ArgumentException">
+    /// Thrown if the name cannot be inferred from the delegate.
+    /// </exception>
+    static TaskName GetTaskNameFromDelegate(Delegate @delegate)
+    {
+        MethodInfo method = @delegate.Method;
+
+        // Check for DurableTaskAttribute on the method
+        DurableTaskAttribute? attribute = method.GetCustomAttribute<DurableTaskAttribute>();
+        if (attribute?.Name.Name is not null and not "")
```
Comment (halspang) [inline src/Abstractions/DurableTaskRegistry.Orchestrators.cs]: Same nit here, let's keep it consistent and use `string.IsNullOrEmpty`.

**Review by halspang — COMMENTED** [review]
Mostly nits, but I am concerned about a potential breaking change. If you could assuage my fear there, I think we're basically good to go.

File: src/Abstractions/Entities/TaskEntity.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (cgillum) [reply src/Abstractions/Entities/TaskEntity.cs:156]: No, because using `string` for `TState` doesn't actually work today. If you try it, it will fail with an obscure error message. I added this so that you get a friendly error message instead of the obscure error message.

File: src/Abstractions/DurableTaskRegistry.Activities.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cgillum) [reply src/Abstractions/DurableTaskRegistry.Activities.cs]: I'll make this change. I think the core issue is that targeting `netstandard2.0` means that we don't have sufficient compiler hints to prove that this value is never null, so we have to do redundant null checks to make the compiler happy.

File: src/Abstractions/DurableTaskRegistry.Orchestrators.cs [hunk 5]
```diff
@@ -220,4 +232,171 @@ public DurableTaskRegistry AddOrchestratorFunc(TaskName name, Action<TaskOrchest
...
+        if (string.IsNullOrEmpty(methodName) || methodName.StartsWith("<", StringComparison.Ordinal))
+        {
+            throw new ArgumentException(
+                "Cannot infer orchestrator name from the delegate. The delegate must either have a " +
+                "[DurableTask] attribute with a name, or be a named method (not a lambda or anonymous delegate).",
+                nameof(@delegate));
+        }
+
+        return new TaskName(methodName);
+    }
```
[the remark this answers is not in the document]
Comment (cgillum) [reply src/Abstractions/DurableTaskRegistry.Orchestrators.cs]: I've moved this method into the common partial class file and updated all calls to GetActivityNameFromDelegate to instead use this method.
